server/player/src/packet_handlers/placement.rs

```diff
@@ -24,66 +24,102 @@ pub fn handle_player_block_placement(
+    packet: PlayerBlockPlacement,
+) {
+    let gamemode = *world.get::<Gamemode>(player);
 
-            let held_item = world.get::<HeldItem>(player).0;
-            let inventory = world.get::<Inventory>(player);
+    let inventory = world.get::<Inventory>(player);
 
-            // Update player's inventory if in survival
-            if gamemode == Gamemode::Survival {
-                if item.amount == 0 {
-                    drop(inventory);
-                    game.disconnect(
-                        player,
-                        world,
-                        "attempted to place block with zero-sized item stack",
-                    );
-                    return;
-                }
+    let item = match inventory.item_in_main_hand(player, world) {
+        Some(item) => item,
+        None => return, // No block to place
+    };
 
-                let item = ItemStack::new(item.ty, item.amount - 1);
-                inventory
-                    .set_item_at(Area::Hotbar, held_item, item)
-                    .unwrap();
+    drop(inventory);
 
-                let event = InventoryUpdateEvent {
-                    slots: smallvec![slot(Area::Hotbar, held_item)],
-                    player,
-                };
-                drop(inventory);
-                game.handle(world, event);
-            }
-        });
+    let block = match item.ty.to_block() {
+        Some(block) => block,
+        None => return, // Item is not a block
+    };
+
+    // TODO: waterlogged blocks, more
+    let pos = match target_block_kind {
+        BlockKind::Grass | BlockKind::TallGrass | BlockKind::Water | BlockKind::Lava => {
+            packet.location
+        }
+        _ => packet.location + packet.face.placement_offset(),
+    };
+
+    game.set_block_at(world, pos, block, BlockUpdateCause::Entity(player));
+
+    let held_item = world.get::<HeldItem>(player).0;
+    let inventory = world.get::<Inventory>(player);
+
+    // Update player's inventory if in survival
+    if gamemode == Gamemode::Survival {
+        if item.amount == 0 {
+            drop(inventory);
+            game.disconnect(
+                player,
+                world,
+                "attempted to place block with zero-sized item stack",
+            );
+            return;
+        }
+
+        let item = ItemStack::new(item.ty, item.amount - 1);
+        inventory
+            .set_item_at(Area::Hotbar, held_item, item)
+            .unwrap();
+
+        let event = InventoryUpdateEvent {
+            slots: smallvec![slot(Area::Hotbar, held_item)],
+            player,
+        };
+        drop(inventory);
+        game.handle(world, event);
+    }
+}
+
+/// Handles interaction with blocks that have an inventory
+pub fn handle_block_interaction(
+    game: &mut Game,
+    world: &mut World,
+    player: Entity,
+    target_block_kind: BlockKind,
+    packet: PlayerBlockPlacement,
+) {
+    let gamemode = *world.get::<Gamemode>(player);
```

Comment:
It might be a good idea to have a global `HashMap<BlockKind, Box<dyn InteractionHandler>>` to handle interaction. Otherwise, this might get messy.

Reply:
Would also get rid of the `Interactable` property requirement